### PR TITLE
4858 connect datadog rum and traces

### PR DESCRIPTION
**Pull request by fungairino**
## What does this PR do?

- Part of https://github.com/pixiebrix/pixiebrix-app/issues/4858
- This change set ensures that we initialize datadog (taking only about 10-15 ms) before we continue in initializing the rest of the UI. This makes it so that the initial requests (ex the `/me` requests) are measured and traced by datadog.

## Demo
Note the datadog headers on the first request.

Screenshot 2024-03-05 at 4 21 17 PM

## Future Work

Check that the tracing is being picked up by the backend app.

## Checklist

- [x] Add tests
- [x] New files added to `src/tsconfig.strictNullChecks.json` (if possible)
- [X] Designate a primary reviewer @twschiller

### CODE REVIEW

File: src/telemetry/performance.ts [hunk 1]
```diff
@@ -92,6 +100,8 @@ export async function initPerformanceMonitoring(): Promise<void> {
     allowFallbackToLocalStorage: true,
   });
 
+  console.log("done init:", Date.now() - start);
```
Comment (fungairino) [inline src/telemetry/performance.ts:103]: woops need to remove this logging

File: src/telemetry/performance.ts [hunk 2]
```diff
@@ -58,13 +59,20 @@ export async function initPerformanceMonitoring(): Promise<void> {
     return;
   }
 
+  const baseUrl = await getBaseURL();
+
+  if (await getDNT()) {
```
Comment (fungairino) [inline src/telemetry/performance.ts:64]: accidental dup, I'll remove this

File: src/telemetry/performance.ts [hunk 3]
```diff
@@ -30,26 +30,25 @@ import {
...
 
 /**
- * Initialize Datadog Real User Monitoring (RUM) for performance monitoring.
+ * Initialize Datadog Real User Monitoring (RUM) for performance monitoring. This should be called once per page load, before
+ * any user interactions or network requests are made.
  */
 export async function initPerformanceMonitoring(): Promise<void> {
+  const environment = process.env.ENVIRONMENT;
+  const applicationId = process.env.DATADOG_APPLICATION_ID;
+  const clientToken = process.env.DATADOG_CLIENT_TOKEN;
```
Comment (johnnymetz) [inline src/telemetry/performance.ts:42]: Just curios, you moving these env vars inside the func to make testing/mocking easier?

Comment (fungairino) [reply src/telemetry/performance.ts:42]: yes!

File: src/telemetry/performance.ts [hunk 4]
```diff
@@ -58,12 +57,17 @@ export async function initPerformanceMonitoring(): Promise<void> {
...
   }
 
-  if (!applicationId || !clientToken) {
-    console.warn("Datadog application ID or client token not configured");
+  const { version_name } = browser.runtime.getManifest();
+
+  if (!applicationId || !clientToken || !version_name || !environment) {
+    console.warn(
+      "Required environment variables for initializing Datadog missing:",
+      { applicationId, clientToken, version_name, environment },
```
Comment (johnnymetz) [inline src/telemetry/performance.ts:65]: Do we want to be logging `clientToken` to the console? Maybe just log the env names, or better yet, the missing env names.

Comment (johnnymetz) [inline src/telemetry/performance.ts:65]: We're logging the env values here. Maybe just log the env names, or better yet, the missing env names.

**Review by johnnymetz — APPROVED** [review]
Few nits but looks great

File: src/telemetry/performance.ts (hunk 4 above)
Comment (fungairino) [reply src/telemetry/performance.ts:65]: I'd assume that these variables shouldn't be confidential and we could log them, but I can instead just log whether they are defined

**Pull request merged**